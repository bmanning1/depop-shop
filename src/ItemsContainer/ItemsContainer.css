.items-body {
    display: flex;
    flex-wrap: wrap;
}

.items-header {
    display: flex;
    flex-direction: column;
    margin: 0.5rem;
}

.results-number {
    margin: 0;
}

.items-container {
    margin: 2rem auto;
}

@media (min-width: 1024px) {
    .items-container {
        margin: 2rem 7rem;
    }

    .filter-buttons button:hover {
        border-color: rgb(215, 215, 215);
        color: white;
        background-color: rgb(215, 215, 215);
    }
}

@media (max-width: 480px) {
    .filter-buttons {
        display: flex;
        flex-wrap: wrap;
        justify-content: center;
    }

    .results-number {
        text-align: center;
    }

    .items-body {
        justify-content: center;
    }
}

.filter-buttons button {
    border: rgb(215, 215, 215) solid 1px;
    background: none;
    font-size: 0.8rem;
    font-weight: 600;
    padding: 0.3rem 1rem;
    cursor: pointer;
    border-radius: 0.2rem;
    width: fit-content;
    margin-top: 1rem;
    transition: 0.3s;
    margin-right: 0.5rem;
}

.filter-buttons button:focus {
    outline: none
}

.filter-buttons button.active {
    border-color: rgb(160, 160, 160);
    color: white;
    background-color: rgb(160, 160, 160);
}

.loader {
    margin: 3rem auto 3rem;
    text-align: center;
}

.loader div {
    width: 0.8rem;
    height: 0.8rem;
    margin: 0.1rem;
    border-radius: 100%;
    background-color: red;
    display: inline-block;
    animation: grow 1.4s infinite ease-in-out;
}

.loader div:nth-of-type(1) {
    animation-delay: -0.32s;
}

.loader div:nth-of-type(2) {
    animation-delay: -0.16s;
}

@keyframes grow {
    0%, 100% {
        transform: scale(0);
    } 50% {
        transform: scale(1);
    }
}

