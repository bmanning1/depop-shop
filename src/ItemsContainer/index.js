import React, { useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import Item from '../Item';
import './ItemsContainer.css';

const Loading = () => (
    <div className="loader">
        <div></div>
        <div></div>
        <div></div>
    </div>
);

export const FilterButton = ({ filters, filter, toggleFilter, label }) => (
    <button className={filters.includes(filter) ? 'active' : ''} onClick={() => toggleFilter(filter)}>
        {label}
    </button>
);

const ItemsContainer = ({ items, loading, toggleLike }) => {
    const [ filters, setFilters ] = useState([]);

    const filteredItems = useMemo(() => {
        return items.filter(item => !(
            (filters.includes('sold') && item.sold !== false) ||
            (filters.includes('liked') && item.liked === false)
        ));
    }, [filters, items]);

    const addFilter = (filter) => setFilters([ ...filters, filter ]);

    const removeFilter = (filterToRemove) => setFilters(filters.filter(filter => filter !== filterToRemove));

    const toggleFilter = (filter) => filters.includes(filter) ? removeFilter(filter) : addFilter(filter);

    const length = loading ? 0 : filteredItems.length;

    return (
        <div className="items-container">
            <div className="items-header">
                <h4 className="results-number">
                    {`${length} Result${length === 1 ? '' : 's'}`}
                </h4>
                <div className="filter-buttons">
                    <FilterButton
                        filters={filters}
                        filter="sold"
                        toggleFilter={toggleFilter}
                        label="Hide Sold Items"
                    />
                    <FilterButton
                        filters={filters}
                        filter="liked"
                        toggleFilter={toggleFilter}
                        label="Show Liked Items"
                    />
                </div>
            </div>
            <div className="items-body">
                {loading
                    ? <Loading />
                    : filteredItems.map(item => <Item key={item.id} toggleLike={toggleLike} {...item}/>)}
            </div>
        </div>
    )
};

ItemsContainer.propTypes = {
    items: PropTypes.arrayOf(PropTypes.exact({
        id: PropTypes.number.isRequired,
        img: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
        brand: PropTypes.string,
        size: PropTypes.string,
        seller: PropTypes.string.isRequired,
        sold: PropTypes.bool.isRequired,
        liked: PropTypes.bool.isRequired
    })),
    loading: PropTypes.bool.isRequired,
    toggleLike: PropTypes.func.isRequired
}

ItemsContainer.defaultProps = {
    items: []
}

export default ItemsContainer;