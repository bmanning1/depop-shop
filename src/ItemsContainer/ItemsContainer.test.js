import React from 'react';
import { shallow } from 'enzyme';
import ItemsContainer, { FilterButton } from '.';

describe('ItemsContainer:', () => {
    let baseComponent; 
    const items = [
        {
            id: 1,
            img: 'http://www.img1Url.com',
            title: 'Jumper',
            price: 35.50,
            brand: 'Supreme',
            size: 'L',
            seller: '@user1',
            sold: false,
            liked: true
        },
        {
            id: 2,
            img: 'http://www.img2Url.com',
            title: 'Trainers',
            price: 25,
            brand: 'Converse',
            seller: '@user2',
            sold: true,
            liked: false
        },
        {
            id: 3,
            img: 'http://www.img3Url.com',
            title: 'Skirt',
            price: 10,
            size: 'M',
            seller: '@user3',
            sold: false,
            liked: false
        },
    ];

    beforeEach(() => {
        baseComponent = shallow(
            <ItemsContainer items={items} loading={false} toggleLike={jest.fn()} />
        );
    });

    test('should match snapshot', () => {
        expect(baseComponent).toMatchSnapshot();
    });

    test('should show the Loading component and 0 results if loading', () => {
        const component = shallow(
            <ItemsContainer items={items} loading={true} toggleLike={jest.fn()} />
        );
        expect(baseComponent).toMatchDiffSnapshot(component);
    });

    test('should handle no items', () => {
        const component = shallow(
            <ItemsContainer items={[]} loading={false} toggleLike={jest.fn()} />
        );
        expect(baseComponent).toMatchDiffSnapshot(component);
    });

    test('should display the correct grammer if there is 1 item', () => {
        const component = shallow(
            <ItemsContainer items={[items[0]]} loading={false} toggleLike={jest.fn()} />
        );
        expect(baseComponent).toMatchDiffSnapshot(component);
    });

    test.each([ ['liked'], ['sold'] ])
        ('should filter by %p items when the respective filter button is clicked', (filter) => {
            const component = shallow(
                <ItemsContainer items={items} loading={false} toggleLike={jest.fn()} />
            );
            component.find(`FilterButton[filter="${filter}"]`).dive().find('button').simulate('click');
            expect(baseComponent).toMatchDiffSnapshot(component);
        }
    );

    test('should handle multiple filters', () => {
            const component = shallow(
                <ItemsContainer items={items} loading={false} toggleLike={jest.fn()} />
            );
            component.find(`FilterButton[filter="sold"]`).dive().find('button').simulate('click');
            component.find('FilterButton[filter="liked"]').dive().find('button').simulate('click');
            expect(baseComponent).toMatchDiffSnapshot(component);
        }
    );
});

describe('FilterButton:', () => {
    let baseComponent, toggleFilter;
    const filter = 'sold'; 

    beforeEach(() => {
        toggleFilter = jest.fn();
        baseComponent = shallow(
            <FilterButton filters={[]} filter={filter} toggleFilter={toggleFilter} label="Hide Sold Items" />
        );
    });

    test('should match snapshot', () => {
        expect(baseComponent).toMatchSnapshot();
    });

    test('should add active styling if the filter is on', () => {
        const component = shallow(
            <FilterButton filters={[filter]} filter={filter} toggleFilter={toggleFilter} label="Hide Sold Items" />
        );
        expect(baseComponent).toMatchDiffSnapshot(component);
    });

    test('should call toggleFilter with the given filter on click', () => {
        baseComponent.simulate('click');
        expect(toggleFilter).toHaveBeenCalledWith(filter);
    });
});
