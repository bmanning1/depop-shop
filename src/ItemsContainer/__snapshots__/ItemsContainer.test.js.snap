// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButton: should add active styling if the filter is on 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <button
-   className=\\"\\"
+   className=\\"active\\"
    onClick={[Function onClick]}
  >
    Hide Sold Items
  </button>"
`;

exports[`FilterButton: should match snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
>
  Hide Sold Items
</button>
`;

exports[`ItemsContainer: should display the correct grammer if there is 1 item 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,11 +5,11 @@
      className=\\"items-header\\"
    >
      <h4
        className=\\"results-number\\"
      >
-       3 Results
+       1 Result
      </h4>
      <div
        className=\\"filter-buttons\\"
      >
        <FilterButton
@@ -38,35 +38,9 @@
        price={35.5}
        seller=\\"@user1\\"
        size=\\"L\\"
        sold={false}
        title=\\"Jumper\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand=\\"Converse\\"
-       id={2}
-       img=\\"http://www.img2Url.com\\"
-       key=\\"2\\"
-       liked={false}
-       price={25}
-       seller=\\"@user2\\"
-       size={null}
-       sold={true}
-       title=\\"Trainers\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand={null}
-       id={3}
-       img=\\"http://www.img3Url.com\\"
-       key=\\"3\\"
-       liked={false}
-       price={10}
-       seller=\\"@user3\\"
-       size=\\"M\\"
-       sold={false}
-       title=\\"Skirt\\"
        toggleLike={[Function mockConstructor]}
      />
    </div>
  </div>"
`;

exports[`ItemsContainer: should filter by "liked" items when the respective filter button is clicked 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,24 +5,32 @@
      className=\\"items-header\\"
    >
      <h4
        className=\\"results-number\\"
      >
-       3 Results
+       1 Result
      </h4>
      <div
        className=\\"filter-buttons\\"
      >
        <FilterButton
          filter=\\"sold\\"
-         filters={Array []}
+         filters={
+           Array [
+             \\"liked\\",
+           ]
+         }
          label=\\"Hide Sold Items\\"
          toggleFilter={[Function toggleFilter]}
        />
        <FilterButton
          filter=\\"liked\\"
-         filters={Array []}
+         filters={
+           Array [
+             \\"liked\\",
+           ]
+         }
          label=\\"Show Liked Items\\"
          toggleFilter={[Function toggleFilter]}
        />
      </div>
    </div>
@@ -38,35 +46,9 @@
        price={35.5}
        seller=\\"@user1\\"
        size=\\"L\\"
        sold={false}
        title=\\"Jumper\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand=\\"Converse\\"
-       id={2}
-       img=\\"http://www.img2Url.com\\"
-       key=\\"2\\"
-       liked={false}
-       price={25}
-       seller=\\"@user2\\"
-       size={null}
-       sold={true}
-       title=\\"Trainers\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand={null}
-       id={3}
-       img=\\"http://www.img3Url.com\\"
-       key=\\"3\\"
-       liked={false}
-       price={10}
-       seller=\\"@user3\\"
-       size=\\"M\\"
-       sold={false}
-       title=\\"Skirt\\"
        toggleLike={[Function mockConstructor]}
      />
    </div>
  </div>"
`;

exports[`ItemsContainer: should filter by "sold" items when the respective filter button is clicked 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,24 +5,32 @@
      className=\\"items-header\\"
    >
      <h4
        className=\\"results-number\\"
      >
-       3 Results
+       2 Results
      </h4>
      <div
        className=\\"filter-buttons\\"
      >
        <FilterButton
          filter=\\"sold\\"
-         filters={Array []}
+         filters={
+           Array [
+             \\"sold\\",
+           ]
+         }
          label=\\"Hide Sold Items\\"
          toggleFilter={[Function toggleFilter]}
        />
        <FilterButton
          filter=\\"liked\\"
-         filters={Array []}
+         filters={
+           Array [
+             \\"sold\\",
+           ]
+         }
          label=\\"Show Liked Items\\"
          toggleFilter={[Function toggleFilter]}
        />
      </div>
    </div>
@@ -38,23 +46,10 @@
        price={35.5}
        seller=\\"@user1\\"
        size=\\"L\\"
        sold={false}
        title=\\"Jumper\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand=\\"Converse\\"
-       id={2}
-       img=\\"http://www.img2Url.com\\"
-       key=\\"2\\"
-       liked={false}
-       price={25}
-       seller=\\"@user2\\"
-       size={null}
-       sold={true}
-       title=\\"Trainers\\"
        toggleLike={[Function mockConstructor]}
      />
      <Item
        brand={null}
        id={3}"
`;

exports[`ItemsContainer: should handle multiple filters 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,24 +5,34 @@
      className=\\"items-header\\"
    >
      <h4
        className=\\"results-number\\"
      >
-       3 Results
+       1 Result
      </h4>
      <div
        className=\\"filter-buttons\\"
      >
        <FilterButton
          filter=\\"sold\\"
-         filters={Array []}
+         filters={
+           Array [
+             \\"sold\\",
+             \\"liked\\",
+           ]
+         }
          label=\\"Hide Sold Items\\"
          toggleFilter={[Function toggleFilter]}
        />
        <FilterButton
          filter=\\"liked\\"
-         filters={Array []}
+         filters={
+           Array [
+             \\"sold\\",
+             \\"liked\\",
+           ]
+         }
          label=\\"Show Liked Items\\"
          toggleFilter={[Function toggleFilter]}
        />
      </div>
    </div>
@@ -38,35 +48,9 @@
        price={35.5}
        seller=\\"@user1\\"
        size=\\"L\\"
        sold={false}
        title=\\"Jumper\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand=\\"Converse\\"
-       id={2}
-       img=\\"http://www.img2Url.com\\"
-       key=\\"2\\"
-       liked={false}
-       price={25}
-       seller=\\"@user2\\"
-       size={null}
-       sold={true}
-       title=\\"Trainers\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand={null}
-       id={3}
-       img=\\"http://www.img3Url.com\\"
-       key=\\"3\\"
-       liked={false}
-       price={10}
-       seller=\\"@user3\\"
-       size=\\"M\\"
-       sold={false}
-       title=\\"Skirt\\"
        toggleLike={[Function mockConstructor]}
      />
    </div>
  </div>"
`;

exports[`ItemsContainer: should handle no items 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,11 +5,11 @@
      className=\\"items-header\\"
    >
      <h4
        className=\\"results-number\\"
      >
-       3 Results
+       0 Results
      </h4>
      <div
        className=\\"filter-buttons\\"
      >
        <FilterButton
@@ -26,47 +26,7 @@
        />
      </div>
    </div>
    <div
      className=\\"items-body\\"
-   >
-     <Item
-       brand=\\"Supreme\\"
-       id={1}
-       img=\\"http://www.img1Url.com\\"
-       key=\\"1\\"
-       liked={true}
-       price={35.5}
-       seller=\\"@user1\\"
-       size=\\"L\\"
-       sold={false}
-       title=\\"Jumper\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand=\\"Converse\\"
-       id={2}
-       img=\\"http://www.img2Url.com\\"
-       key=\\"2\\"
-       liked={false}
-       price={25}
-       seller=\\"@user2\\"
-       size={null}
-       sold={true}
-       title=\\"Trainers\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand={null}
-       id={3}
-       img=\\"http://www.img3Url.com\\"
-       key=\\"3\\"
-       liked={false}
-       price={10}
-       seller=\\"@user3\\"
-       size=\\"M\\"
-       sold={false}
-       title=\\"Skirt\\"
-       toggleLike={[Function mockConstructor]}
    />
-   </div>
  </div>"
`;

exports[`ItemsContainer: should match snapshot 1`] = `
<div
  className="items-container"
>
  <div
    className="items-header"
  >
    <h4
      className="results-number"
    >
      3 Results
    </h4>
    <div
      className="filter-buttons"
    >
      <FilterButton
        filter="sold"
        filters={Array []}
        label="Hide Sold Items"
        toggleFilter={[Function]}
      />
      <FilterButton
        filter="liked"
        filters={Array []}
        label="Show Liked Items"
        toggleFilter={[Function]}
      />
    </div>
  </div>
  <div
    className="items-body"
  >
    <Item
      brand="Supreme"
      id={1}
      img="http://www.img1Url.com"
      key="1"
      liked={true}
      price={35.5}
      seller="@user1"
      size="L"
      sold={false}
      title="Jumper"
      toggleLike={[MockFunction]}
    />
    <Item
      brand="Converse"
      id={2}
      img="http://www.img2Url.com"
      key="2"
      liked={false}
      price={25}
      seller="@user2"
      size={null}
      sold={true}
      title="Trainers"
      toggleLike={[MockFunction]}
    />
    <Item
      brand={null}
      id={3}
      img="http://www.img3Url.com"
      key="3"
      liked={false}
      price={10}
      seller="@user3"
      size="M"
      sold={false}
      title="Skirt"
      toggleLike={[MockFunction]}
    />
  </div>
</div>
`;

exports[`ItemsContainer: should show the Loading component and 0 results if loading 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,11 +5,11 @@
      className=\\"items-header\\"
    >
      <h4
        className=\\"results-number\\"
      >
-       3 Results
+       0 Results
      </h4>
      <div
        className=\\"filter-buttons\\"
      >
        <FilterButton
@@ -27,46 +27,8 @@
      </div>
    </div>
    <div
      className=\\"items-body\\"
    >
-     <Item
-       brand=\\"Supreme\\"
-       id={1}
-       img=\\"http://www.img1Url.com\\"
-       key=\\"1\\"
-       liked={true}
-       price={35.5}
-       seller=\\"@user1\\"
-       size=\\"L\\"
-       sold={false}
-       title=\\"Jumper\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand=\\"Converse\\"
-       id={2}
-       img=\\"http://www.img2Url.com\\"
-       key=\\"2\\"
-       liked={false}
-       price={25}
-       seller=\\"@user2\\"
-       size={null}
-       sold={true}
-       title=\\"Trainers\\"
-       toggleLike={[Function mockConstructor]}
-     />
-     <Item
-       brand={null}
-       id={3}
-       img=\\"http://www.img3Url.com\\"
-       key=\\"3\\"
-       liked={false}
-       price={10}
-       seller=\\"@user3\\"
-       size=\\"M\\"
-       sold={false}
-       title=\\"Skirt\\"
-       toggleLike={[Function mockConstructor]}
-     />
+     <Loading />
    </div>
  </div>"
`;
